sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,152 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"path"
+
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/inotify"
+)
+
+type inotifyWatcher struct {
+	wd        *inotify.Watcher
+	addWatch  chan *groupToWatch
+	groupList []*groupToWatch
+	pathList  map[string]int32
+}
+
+type groupToWatch struct {
+	name    string
+	folders []string
+	channel chan string
+}
+
+var currentWatcher *inotifyWatcher = &inotifyWatcher{
+	wd:       nil,
+	pathList: make(map[string]int32),
+	addWatch: make(chan *groupToWatch),
+}
+
+func addWatch(newWatch *groupToWatch) {
+	var folderList []string
+	for _, fullPath := range newWatch.folders {
+		// It's not possible to use inotify.InDeleteSelf in /sys/fs because it
+		// isn't triggered, so we must monitor the parent folder and use InDelete
+		basePath := path.Dir(fullPath)
+		if _, exists := currentWatcher.pathList[basePath]; !exists {
+			currentWatcher.pathList[basePath] = 0
+			if err := currentWatcher.wd.AddWatch(basePath, inotify.InDelete); err != nil {
+				delete(currentWatcher.pathList, basePath)
+				continue
+			}
+		}
+		currentWatcher.pathList[fullPath]++
+		if osutil.FileExists(fullPath) {
+			folderList = append(folderList, fullPath)
+		} else {
+			removePath(fullPath)
+		}
+	}
+	if len(folderList) == 0 {
+		newWatch.channel <- newWatch.name
+	} else {
+		newWatch.folders = folderList
+		currentWatcher.groupList = append(currentWatcher.groupList, newWatch)
+	}
+}
+
+func removePath(fullPath string) {
+	currentWatcher.pathList[fullPath]--
+	if currentWatcher.pathList[fullPath] == 0 {
+		currentWatcher.wd.RemoveWatch(fullPath)
+		delete(currentWatcher.pathList, fullPath)
+	}
+}
+
+func processEvent(watch *groupToWatch, event *inotify.Event) bool {
+	var tmpFolders []string
+	for _, fullPath := range watch.folders {
+		if fullPath != event.Name {
+			tmpFolders = append(tmpFolders, fullPath)
+		} else {
+			removePath(fullPath)
+		}
+	}
+	watch.folders = tmpFolders
+	if len(tmpFolders) == 0 {
+		watch.channel <- watch.name
+		return false
+	}
+	return true
+}
+
+func watcherMainLoop() {
```

Comment:
Consider making this function into a method on `inotifyWatcher`. Same for a few of the above functions.

Reply:
Done.